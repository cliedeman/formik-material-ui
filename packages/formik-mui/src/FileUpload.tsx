import * as React from 'react';
import { ErrorMessage, FieldProps, getIn } from 'formik';
import Button, { ButtonProps } from '@mui/material/Button';
import FormControl from '@mui/material/FormControl';
import FormHelperText from '@mui/material/FormHelperText';
import InputAdornment from '@mui/material/InputAdornment';
import TextField from '@mui/material/TextField';
import { InputProps } from '@mui/material/Input';

export interface FileUploadProps extends FieldProps {
  label: string;
  accept: string;
  disabled?: boolean;
  fullWidth?: boolean;
  margin?: 'dense' | 'normal' | 'none';
  uploadButtonText?: string;
  noFileChosenText?: string;
  InputProps?: Omit<
    InputProps,
    'name' | 'type' | 'label' | 'readOnly' | 'endAdornment' | 'onClick'
  >;
  ButtonProps?: Omit<ButtonProps, 'onClick'>;
}

export const FileUpload = ({
  field,
  form: { isSubmitting, touched, errors, setFieldValue },
  label,
  accept,
  disabled = false,
  fullWidth,
  margin = 'none',
  uploadButtonText = 'Choose file',
  noFileChosenText = 'No file chosen',
  InputProps: inputProps,
  ButtonProps: buttonProps,
  ...props
}: FileUploadProps) => {
  const hiddenFileInputRef = React.useRef<HTMLInputElement>(null);

  const error = getIn(touched, field.name) && getIn(errors, field.name);

  const handleUploadClick = () => {
    hiddenFileInputRef.current?.click();
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleFileUpload = (event: React.ChangeEvent<any>) => {
    if (inputProps?.onChange) {
      inputProps.onChange(event);
    } else {
      const file = event.currentTarget.files[0];
      setFieldValue(field.name, file);
    }
  };

  return (
    <FormControl fullWidth={fullWidth} margin={margin}>
      <input
        ref={hiddenFileInputRef}
        style={{ display: 'none' }}
        onChange={handleFileUpload}
        type="file"
        accept={accept}
      />
      <TextField
        error={!!error}
        type="text"
        value={field.value ? field.value.name : noFileChosenText}
        label={label}
        InputProps={{
          disabled: disabled || isSubmitting,
          ...inputProps,
          onClick: handleUploadClick,
          readOnly: true,
          endAdornment: (
            <InputAdornment position="end">
              <Button
                disabled={disabled || isSubmitting}
                color="secondary"
                disableElevation
                {...buttonProps}
                onClick={handleUploadClick}
              >
                {uploadButtonText}
              </Button>
            </InputAdornment>
          ),
        }}
        {...props}
      />
      <ErrorMessage
        name={field.name}
        render={(error) => <FormHelperText error>{error}</FormHelperText>}
      />
    </FormControl>
  );
};
