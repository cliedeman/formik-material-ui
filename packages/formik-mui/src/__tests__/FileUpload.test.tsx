import * as React from 'react';
import { test, expect } from '@jest/globals';
import { Formik, Form, Field } from 'formik';
import renderer from 'react-test-renderer';

import { FileUpload } from '../FileUpload';

test('FileUpload Renders Correctly', () => {
  const component = renderer.create(
    <Formik onSubmit={() => {}} initialValues={{ test: '' }}>
      <Form>
        <Field component={FileUpload} name="test" label="Select" />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
