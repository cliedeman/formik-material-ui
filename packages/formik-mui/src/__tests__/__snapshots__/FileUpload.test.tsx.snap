// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload Renders Correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
      >
        Select
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          readOnly={true}
          required={false}
          type="text"
          value="No file chosen"
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            className="MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root css-1jkbrr3-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            Choose file
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-186xcr5"
          >
            <span>
              Select
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</form>
`;
